src/network/mining_pool.rs

```diff
@@ -0,0 +1,324 @@
+                            let valid_owner = coinbase_records
+                                .iter()
+                                .map(|r| Some(r.owner()) == self.mining_pool_address)
+                                .fold(false, |a, b| a || b);
+
+                            if !valid_owner {
+                                warn!("[ProposedBlock] Peer {} sent a candidate block with an invalid owner.", peer_ip);
+                                return;
+                            }
+
+                            coinbase_records
+                        }
+                        Err(err) => {
+                            warn!("[ProposedBlock] {}", err);
+                            return;
+                        }
+                    };
+
+                    // Determine the score to add for the miner.
+                    let proof_bytes = match block.header().proof() {
+                        Some(proof) => match proof.to_bytes_le() {
+                            Ok(bytes) => bytes,
+                            Err(err) => {
+                                warn!("[ProposedBlock] {}", err);
+                                return;
+                            }
+                        },
+                        None => {
+                            warn!("[ProposedBlock] Peer {} sent a candidate block with a missing proof.", peer_ip);
+                            return;
+                        }
+                    };
+
+                    let hash_difficulty = sha256d_to_u64(&proof_bytes);
+                    let mut info = self.miner_info.write().await;
+                    let share_difficulty = match info.get(&miner_address) {
+                        Some((_, share_difficulty, _)) => *share_difficulty,
+                        None => {
+                            let share_difficulty = current_template.difficulty_target.saturating_mul(50);
+                            self.miner_info
+                                .write()
+                                .await
+                                .insert(miner_address, (chrono::Utc::now().timestamp(), share_difficulty, 0));
+
+                            share_difficulty
+                        }
+                    };
+
+                    if hash_difficulty > share_difficulty {
+                        warn!("[ProposedBlock] faulty share submitted by {}", miner_address);
+                        return;
+                    }
+
+                    // Update the score for the miner.
+                    // TODO: add round stuff
+                    if let Err(error) = self.state.add_shares(block.height(), &miner_address, 1) {
+                        warn!("[ProposedBlock] {}", error);
+                    }
+
+                    // Update miner info for this miner.
+                    let mut miner_info = *info.get(&miner_address).expect("miner should have existing info");
```

Comment:
`miner_info` can easily be mistaken for `info` (`self.miner_info` locked for writing) here

Reply:
also, why is this not `get_mut`?